mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h

```diff
@@ -134,7 +134,77 @@ struct DerivedBlockGemmParams {
   int64_t gemmNCuwavesPerBlock;
 };
 
-class PopulateParams {
+template <typename InitParamType> class BasePopulateParams {
+private:
+  struct InitParamData {
+    InitParamType paramSet;
+    size_t original_pos;
+    int64_t padding_amount;
+
+    bool operator<(const InitParamData &rhs) {
+      if (this->padding_amount < rhs.padding_amount) {
+        return true;
+      } else if (this->padding_amount == rhs.padding_amount) {
+        if (this->original_pos < rhs.original_pos) {
+          return true;
+        }
+        return false;
+      }
+      return false;
+    }
+  };
+
+  std::vector<InitParamData>
+  createParamData(const ArrayRef<InitParamType> &initParams,
+                  const GemmSize &gemmSize) {
+    std::vector<InitParamData> res;
+    for (size_t pos = 0; pos < initParams.size(); pos++) {
+      InitParamType paramSet = initParams[pos];
+      InitParamData paramData;
+      paramData.paramSet = paramSet;
+      paramData.original_pos = pos;
+      paramData.padding_amount = calculatePaddingAmount(paramSet, gemmSize);
+      assert(paramData.original_pos >= 0);
+      assert(paramData.padding_amount >= 0);
+      res.push_back(paramData);
+    }
+    return res;
+  }
+
+protected:
+  // Interface function to check whether the given GEMM is valid
+  virtual LogicalResult isValidGemm(const InitParamType &param,
+                                    const GemmSize &gemmSize) const = 0;
+
+  // Interface function to calculate padding amount of a given param set
+  virtual int64_t calculatePaddingAmount(const InitParamType &params,
+                                         const GemmSize &gemmSize) const = 0;
+
+  // This function will order the initParams used in a non-tuning flow to
+  // prioritize params that does not require padding to come first while
+  // otherwise maintaining the provided order.
+  std::vector<InitParamType>
+  orderInitParams(const std::vector<InitParamType> &initParams,
```

Comment:
`const std::vector<T> &` is spelled `ArrayRef<T>`

Reply:
Done